laurat007/assignment3/Word.java

```diff
@@ -0,0 +1,113 @@
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ *
+ * @author Laura
+ * 
+ *  This program returns the words that can be formed using the letters in a 
+ * given grid by moving left, right, up, down and on diagonals.
+ *  Moving in the grid is similar to a backtracking from each Cell which stops
+ * if the current word formed is not a prefix, and if it finds a word it is added
+ * to solution.
+ * 
+ */
+
+public class Word {
+    
+    static TreeSet<String> words(Grid G) {
+        TreeSet<String> solution = new TreeSet<>();
+        boolean[][] visited = new boolean[G.rows][G.columns];
+        
+        // start search from each existing cell
+        for(int i = 0; i < G.rows; i++) {
+            for(int j = 0; j < G.columns; j++) {
+                backtrack(i, j, G, solution, 
+                    Character.toString(G.grid[i][j]), visited);
+                visited[i][j] = false;
+            }
+        }
+        return solution;
+    }
+
+    static private void backtrack(int startingCellX, int startingCellY, Grid G, 
+        TreeSet<String> solution, String word, boolean[][] visited) {
+        
+        visited[startingCellX][startingCellY] = true;
+        
+        // stop condition
+        if(! G.dictionary.isPrefix(word)) {
+            return;
+        }
+        // if solution found, will be added
+        if(G.dictionary.isWord(word))
+            solution.add(word);
+        
+        // try moving
```

Comment:
You have eight calls that essentially do the same thing (code duplication) apart from the directions they call backtrack on.
You could define a list of deltas (or two lists) and iterate through them and get rid of the code duplication. 
Let me know if this is unclear and you need help with it :)

Reply:
I am not sure if this is what you meant, but this is what I found trying to figure it out.